### DISCUSSION TITLE
403 errors only for certain users

### DISCUSSION THREAD

**Discussion by scouech** [directus/directus#5424] [Archive]
Hi! 👋 

I have a role, this role has rights to certain collections and for most users it works, however, now that the quantity of users for this role is increasing I get more and more 403 errors for certain users. 

There is actually no difference between the users, they access the same content within the same role. 
I don't know if Directus has a log, however on nginx in the access log I can spot the 403 calls.

It was working all fine with a few users, but now we have 1000 for that role it seems that is getting more and more frequent.

Directus 61
Postgre
Hosted on DigitalOcean
Nginx

**Nginx.conf**
```
user www-data;
worker_processes 4;
pid /run/nginx.pid;
include /etc/nginx/modules-enabled/*.conf;

events {
	worker_connections 1024;
	multi_accept on;
}

http {

	##
	# Basic Settings
	##

	sendfile on;
	tcp_nopush on;
	tcp_nodelay on;
	keepalive_timeout 90;
	send_timeout 120;
	types_hash_max_size 2048;
	reset_timedout_connection on;
	# server_tokens off;

	# server_names_hash_bucket_size 64;
	# server_name_in_redirect off;

	include /etc/nginx/mime.types;
	default_type application/octet-stream;

	##
	# SSL Settings
	##

	ssl_protocols TLSv1 TLSv1.1 TLSv1.2 TLSv1.3; # Dropping SSLv3, ref: POODLE
	ssl_prefer_server_ciphers on;

	##
	# Logging Settings
	##

	access_log /var/log/nginx/access.log;
	error_log /var/log/nginx/error.log;

	##
	# Gzip Settings
	##

	gzip on;

	gzip_vary on;
	gzip_proxied any;
	gzip_comp_level 6;
	gzip_buffers 16 8k;
	gzip_min_length 1000;
	gzip_http_version 1.1;
	gzip_types text/plain text/css application/json application/javascript text/xml application/xml application/xml+rss text/javascript;

	##
	# Virtual Host Configs
	##

	include /etc/nginx/conf.d/*.conf;
	include /etc/nginx/sites-enabled/*;

    	##
    	# Override some buffer limitations, will prevent DDOS too.
    	##

	client_body_buffer_size 128K;
	client_header_buffer_size 1k;
	client_max_body_size 10000m;
	large_client_header_buffers 4 16k;

    	##
    	# send the client a "request timed out" if the body is not loaded
    	# by this time. Default 60.
    	##
    	client_body_timeout   32;
    	client_header_timeout 32;
}
```

**extract of site-available config file** + (i'm using certbot)

```
 location / {
        proxy_pass http://127.0.0.1:8055;
        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto $scheme;

	autoindex on;
	autoindex_exact_size off;
    }

```

**Package.json**
```
{
  "name": "XXXXX",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "start": "directus start"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "dependencies": {
    "directus": "^9.0.0-rc.53",
    "pg": "^8.5.1"
  }
}
```

Hope you can help!

**Comment by rijkvanzanten** [comment]
What error is thrown with the 403? Is it a Directus Forbidden error, or could it be on the server level?

Does a single endpoint return this, or does it happen sporadically across multiple endpoints? 

Could it be that those 403 users are trying to hit the API with an invalid token?

**Reply by scouech** [reply]
1. mmmm we don't get anything that says it's from Directus, it just says 403. 
2. It seems to be random endpoints, multiple endpoints and random times. Sometimes it would work perfectly and then not
3. People getting 403 are using a valid token

I guess it could probably be our server? It's really hard to tell.. Just for context it's 1 role with more than 2000 users at the moment, it seems that it was all fine when we had a few hundred and the problem started to appear when the number increased.. (not that it understand the problem this much haha)

**Reply by rijkvanzanten** [reply]
> we don't get anything that says it's from Directus

Directus will only throw 403s with the response body as JSON with an `errors` key with the FORBIDDEN exception in it. If that response body is missing, it must be coming from the server 🤔 

> Just for context it's 1 role with more than 2000 users at the moment

I mean a user is a user and a role is a role, so that shouldn't cause  issues 🤔 Though there could always be a bug somewhere hidden in plain sight of course.. 

Is there any pattern in the types of users where this is happening?

**Reply by scouech** [reply]
You made me think and I contacted one of the users to give me their user id (we have a little hidden feature to show it), I looked for this ID in the list of users for this specific role and couldn't find it. 
I exported the entire user table from postgre directly and discovered that this user (and a few others that have reported a bug) have been created but for some reason the user isn't attached to the right role. 

We are using the rest api and are doing the call properly.. it's really odd that the user is being created without a role despite passing the role id in the header... You would always expect the API to create the user with the right role id if the role id is in the header right?

**Reply by rijkvanzanten** [reply]
> We are using the rest api and are doing the call properly..

How are you doing the call exactly? Are you sure it's properly? 

> passing the role id in the header

In the header?

> You would always expect the API to create the user with the right role id if the role id is in the header right?

Assuming you mean the post body, then yes!

**Reply by scouech** [reply]
sorry , body yes haha! 😆  got confused 

We do 

`POST https://URL/users/`

and in the body:

```
{
"email": "email@email.user",
"password": "PASSWORD",
"role": "97712b9c-db0c-4235-b561-e8b58711d857",
"token": "xxxxxxx",
"device_model": "device model",
"device_os": "device os",
"device_size": "device size",
"device_platform": "device platform",
"device_language": "device language",
"date_joined": "timestamp of when account was created"
}
```

It works via postman and via the app (we have an app).. at least when we test it and for most users... From what I can see from the export at least 100 people don't have the role (it's empty).. 
What I find odd is that all other fields are populated properly, except the role id... 

Is there a way I could default user accounts to a specific role?

**Reply by rijkvanzanten** [reply]
Are you submitting that role from a variable in your own code? Maybe that's undefined at the point you're creating the user? 🤔

**Reply by scouech** [reply]
mmm good question. I will check with Mike the app developer 👨‍💻  it might be it! 
I guess in the meantime I can add the role manually to all the users who don't have one.
Do you think there is a way I can default new users to a specific role? I guess it would require some development?

**Reply by rijkvanzanten** [reply]
You could use a preset on the role that you're using to create these users to pre-set the role. Other than that, you could create a custom hook that force overrides the hook on every user create, but that might be a bit overkill for this usage 👍🏻

**Reply by scouech** [reply]
mmm sorry if I sound stupid, but how do you create a preset like that?

**Reply by rijkvanzanten** [reply]
Under Settings > Roles & Permissions and then the create permission for directus_users has a tab called presets 👍🏻 You can set any default values that only apply to that role in there

**Reply by scouech** [reply]
omg this is really good! So I guess we can remove the role from the post body completely with this? 
wow you did a very good job with Directus, it's really flexible!!

**Reply by rijkvanzanten** [reply]
Yeah you could! You could prevent the registration from setting the role itself altogether with field permissions, and then use the preset to fill it 🙂

**Reply by scouech** [reply]
that's so good!!!! Thank you so much for your help! It should fix the problem perfectly :) !!!

**Comment by scouech** [comment]
@rijkvanzanten  - what we discussed previously didn't actually fix the random 403 errors unfortunately. 
It's really hard to reproduce, but I check the panel and managed to randomly reproduce the error (see screenshots attached). This is the exact same message as some users get when trying to access the content... 
It seems that the errors happens mostly on assets / files.

FYI: We're using cloudflare on top of the droplet, but we deactivated it just in case if it was because of it.

![Screenshot 2021-05-05 at 12 54 28 PM](https://user-images.githubusercontent.com/26603052/117132019-8ace1180-ada2-11eb-84fe-f2b9195900d2.png)

![Screenshot 2021-05-05 at 1 02 10 PM](https://user-images.githubusercontent.com/26603052/117132050-94577980-ada2-11eb-8976-d39883837ab0.png)

![Screenshot 2021-05-05 at 12 54 52 PM](https://user-images.githubusercontent.com/26603052/117132071-9a4d5a80-ada2-11eb-97c5-07ad4e98f12b.png)

![Screenshot 2021-05-05 at 12 54 42 PM](https://user-images.githubusercontent.com/26603052/117132089-a0433b80-ada2-11eb-9bc9-dc78f0e020f8.png)

**Reply by scouech** [reply]
and nginx returns 403 errors too (see below). 

<img width="742" alt="Screenshot 2021-05-05 at 1 10 47 PM" src="https://user-images.githubusercontent.com/26603052/117132583-5149d600-ada3-11eb-8820-20a1c43d0465.png">

**Reply by rijkvanzanten** [reply]
In all of these cases, it looks like the user doesn't have read access to the file 🤔 What are the permissions to directus_files for this user?

**Reply by scouech** [reply]
Here are the permissions:

<img width="805" alt="Screenshot 2021-05-05 at 4 43 11 PM" src="https://user-images.githubusercontent.com/26603052/117160326-4e5dde00-adc1-11eb-8a8e-6e60900ff78d.png">

**Reply by rijkvanzanten** [reply]
Strange! Does it consistently happen to the same file? How are you sending the access token? Could it be that the header is stripped out by a CDN or proxy? 🤔
